docs/src/DeveloperDocumentation/styleguide.md

```diff
@@ -241,3 +217,249 @@ existing block.
+e.g. to print the `base_ring(X)` when in `one line` mode.
+It exists to make sure that `one line` stays compact and human readable.
+
+Top-level REPL printing of an object will use `detailed` mode by default
+```julia
+julia> X
+detailed
+```
+Inside nested structures, e.g. inside a `Vector`, the `one line` mode is used.
+```julia
+julia> [X,X]
+3-element Vector{TypeofX{T}}
+one line
+one line
+one line
+```
+
+##### An Example for the 2 + 1 print modes
+```
+# detailed
+General linear group of degree 24
+  over Finite field of degree 7 over GF(29)
+
+# one line
+General linear group of degree 24 over GF(29^7)
+
+# supercompact
+General linear group
+```
+
+The print modes are specified as follows
+#### Detailed printing
+- the output must make sense as a standalone without context to non-specialists
+- the number of output lines should fit in the terminal
+- if the object is simple enough use only one line
+- use indentation and (usually) `one line` to print substructures
+#### One line printing
+- the output must print in one line
+- should make sense as a standalone without context
+- variable names/generators/relations should not be printed only their number.
+- Only the first word is capitalized e.g. `Polynomial ring`
+- one should use `:supercompact` for nested printing in compact
+- nested calls to `one line` (if you think them really necessary) should be at the end,
+  so that one can read sequentially. Calls to `:supercompact` can be anywhere.
+- commas must be enclosed in brackets so that printing tuples stays unambiguous
+#### Super compact printing
+- a user readable version of the main (mathematical) type.
+- a single term or a symbol/letter mimicking mathematical notation
+- should usually only depend on the type and not of the type parameters or of
+  the concrete instance - exceptions of this rule are possible e.g. for `GF(2)`
+- no nested printing. In particular variable names and `base_ring` must not be displayed.
+  This ensures that `one line` and `:supercompact` stay compact even for complicated things.
+  If you want nested printing use `one line` or `detailed`.
+
+### Implementing show functions
+
+Here is the translation between `:detail`, `one line` and `:supercompact`.
+
+```
+print(io, "text/plain", x)                 # detailed printing
+print(io, x)                               # one line printing
+print(IOContext(:supercompact => true), x) # supercompact printing
+```
+
+For reference, string interpolation `"$(x)"` will also use `print(io, x)`.
+
+#### Mockup
+
+##### Detailed printing with a new line.
+
+```julia
+struct NewRing
+  base_ring
+end
+
+base_ring(R::NewRing) = R.base_ring
+```
+This implements the detailed printing.
+Note that the new line is needed for this to work
+because of how the julia printing system works.
+The last print statement must not add a new line.
+
+```julia
+function Base.show(io::IO, ::MIME"text/plain", R::NewRing)
```

Comment:
@simonbrandhorst suggested that we turn these running examples into doctests, which I think is a good idea. But of course we can do this *after* merging this PR

Reply:
I did not like the `julia>` prompt in the examples. It messes with being able to copy paste it for the new code.